packages/evm/contracts/adapters/pnetwork/PNetworkMessageRelayer.sol

```diff
@@ -0,0 +1,122 @@
+// SPDX-License-Identifier: LGPL-3.0-only
+pragma solidity ^0.8.17;
+
+import { IMessageRelay } from "../../interfaces/IMessageRelay.sol";
+import { IErc20Vault } from "./IErc20Vault.sol";
+import { IPToken } from "./IPToken.sol";
+import { Yaho } from "../../Yaho.sol";
+import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts-upgradeable/utils/introspection/IERC1820RegistryUpgradeable.sol";
+import "@openzeppelin/contracts/token/ERC777/IERC777.sol";
+
+contract PNetworkMessageRelay is IMessageRelay {
+    bytes32 private constant TOKENS_RECIPIENT_INTERFACE_HASH = keccak256("ERC777TokensRecipient");
+    IERC1820RegistryUpgradeable private constant _erc1820 =
+        IERC1820RegistryUpgradeable(0x1820a4B7618BdE71Dce8cdc73aAB6C95905faD24);
+
+    address public immutable token;
+    address public immutable vault;
+    Yaho public immutable yaho;
+    bytes4[] private _supportedNetworkIds;
+
+    uint256 private constant SWAP_AMOUNT = 100;
+
+    event MessageRelayed(address indexed emitter, uint256 indexed messageId);
+    event NetworkIdAdded(bytes4 networkId);
+
+    error AlreadyExistingNetworkId(bytes4 networkId);
+
+    constructor(address _vault, address _token, Yaho _yaho) {
+        vault = _vault;
+        token = _token;
+        yaho = _yaho;
+        _erc1820.setInterfaceImplementer(address(this), TOKENS_RECIPIENT_INTERFACE_HASH, address(this));
+    }
+
+    // TODO: limit access to this function
+    function addNetwork(bytes4 networkId) public {
+        if (_isNetworkSupported(networkId)) revert AlreadyExistingNetworkId(networkId);
+        _supportedNetworkIds.push(networkId);
+        emit NetworkIdAdded(networkId);
+    }
+
+    function relayMessages(
+        uint256[] memory messageIds,
+        address pnetworkAdapter
+    ) public payable returns (bytes32 receipt) {
+        bytes32[] memory hashes = new bytes32[](messageIds.length);
+        for (uint256 i = 0; i < messageIds.length; i++) {
+            uint256 id = messageIds[i];
+            hashes[i] = yaho.hashes(id);
+            emit MessageRelayed(address(this), messageIds[i]);
+        }
+        bytes memory userData = abi.encode(messageIds, hashes, block.chainid);
+        if (vault != address(0)) {
+            for (uint256 index = 0; index < _supportedNetworkIds.length; index++) {
```

Comment:
Can you use a single variable instead of `supportedNetworkIds`? In this way you can set it as immutable within the constructor and remove `addNetwork`. I know that it implicates to the deploy a contract for each network but this is the way for which current message relays work.

Reply:
Done!